### PR TITLE
[Jobs] Report managed job tasks that stalled before they ever ran

### PR DESCRIPTION
**Pull request by DanielZhangQD**
## Summary

A managed job stuck in `PENDING` because of a system failure looks identical from the dashboard to one legitimately waiting for resources. This adds the predicate that tells them apart, and exports it as metrics.

Two phases, each a positive statement about our own scheduler rather than a classification of somebody else's error text — which is what lets them run without an allow-list of errors to report on:

- **`never_claimed`** — nothing has picked the task up at all. It has never been provisioned, so the wait cannot be a resource wait: a resource wait needs somebody already asking for the resource. Excluded: a task whose sibling is progressing, one behind higher priority, and a batch job whose pool is occupied.
- **`unattended`** — something claimed the task and then stopped driving its launch. Suppressed by a live request for its cluster, a launch attempt still working, or a `job_events` row of its own inside 15 minutes.

`sky/jobs/stall.py` answers only *which tasks are in a phase*. The two scans are separate entry points with no wrapper that runs both: they read different stores and must be able to fail and go dormant independently.

`ManagedJobsStallCollector` exports them, **separate from `ManagedJobsCollector`** rather than folded into it. That collector's `_refresh` is all-or-nothing by design and a `ResilientCollector` can neither time out nor cancel a refresh in flight, so a wedged stall query there would freeze `sky_managed_jobs_count`, which an existing alert depends on. Inside the new collector, one `try` per phase.

New metrics, all labelled by `phase` (`never_claimed` / `unattended`):

| metric | |
|---|---|
| `sky_managed_jobs_stalled{phase, workspace}` | count |
| `sky_managed_jobs_stall_seconds_max{phase, workspace}` | age of the oldest |
| `sky_managed_jobs_stall_truncated{phase}` | 1 when the count is a floor rather than a total |
| `sky_managed_jobs_stall_scan_timestamp_seconds{phase}` | last scan that completed |

**Zero versus absent.** A phase that ran and found nothing reports `0`; a phase that could not run reports nothing at all. So absence means *not measured*, never *healthy* — the `up`-is-absent-rather-than-0 trap, which has blinded an alert before. The scan-timestamp metric is what lets a rule say that without `absent()`.

**Bounding.** Nothing above a running query can interrupt it, so the bound is set on the database: `SET LOCAL statement_timeout` on the managed-jobs engine, following `_bound_user_upsert_transaction`. It does not bound a whole scan — the other two stores are read unbounded — and the comment says so rather than overstating it. Postgres only; SQLite has no equivalent.

**Thresholds** default to 10 and 15 minutes, overridable through `SKYPILOT_SERVER_`-prefixed variables. That prefix is what keeps a client from supplying them: `request_body_env_vars` forwards `SKYPILOT_` variables except those carrying it.

**Migration 028** adds `ix_spot_unattended`. Partial, following 027: it holds claimed work still in flight and a row leaves it the moment it starts or ends. Its predicate is one constant shared three ways — the index, the migration, and the query's `WHERE` — so they cannot drift. An index whose predicate no longer covers its query is silently not used and nothing fails.

## Test plan

Unit tests, 48 new, each suppression case with its baseline arm beside it — the row that must be reported sits next to the row that must not, and the assertion names both.

```bash
pytest tests/unit_tests/test_sky/jobs/test_stall.py \
       tests/unit_tests/test_sky/server/test_metrics_stall_collector.py \
       tests/unit_tests/test_sky/db/test_timeline_index_usage.py
```

**Mutation-checked.** Fourteen invariants were individually reverted in a throwaway pytest plugin and each was confirmed to turn the suite red: the priority exclusion, the progressing-sibling exclusion, the pool self-exclusion, `status = 'PENDING'`, `submitted_at IS NULL`, `WINDING_DOWN`, the live-request suppressor, the attempt age bound, the witness query's self-exclusion, the threshold validation, the narrowed request projection, the collector's per-phase `try`, its zero emission, and the truncation gauge.

Three passes of this found four tests that survived their own mutation, and all four were rewritten. Two had reached for a module attribute an autouse fixture had already stubbed, so they were asserting against the stub; the real bodies are now called through names captured at import. One collector-isolation test only covered a failure in the *second* phase — which a single shared `try` also survives — so it is now parameterized over both orders.

**Index reachability.** `EXPLAIN QUERY PLAN` asserts the real SQL string from `stall.py` reaches `ix_spot_unattended`, with the reverse control that widening the predicate stops it. SQLite only, which the test docstring says: Postgres decides partial-index implication in its own planner.

**Cost.** 10k rows with 200 tasks in flight: `never_claimed` 0.2 ms, `unattended` 14.6 ms median, against a 30 s refresh interval, off the scrape path on a background thread.

## Manual verification

```bash
# On a consolidation-mode API server with metrics enabled:
curl -s localhost:46580/metrics | grep sky_managed_jobs_stall
# Expect both phases present, including zeros, plus a scan timestamp for each.

# Force one phase to fail and confirm the other keeps reporting:
#   the failing phase's series disappear; the other's keep updating.
```

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/skypilot-org/skypilot/pull/10824" target="_blank"><picture><source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-devin-review-dark.svg?v=4">Devin Review</picture></a>

### CODE REVIEW

File: sky/server/metrics.py [hunk 1]
```diff
@@ -1212,6 +1361,15 @@ def maybe_register_managed_jobs_collector():
         prom.REGISTRY.register(_MANAGED_JOBS_COLLECTOR)
     except ValueError:
         pass
+    # Registered alongside rather than inside, so a wedged stall scan cannot
+    # freeze the gauges above it. Same consolidation guard: both read the
+    # managed-jobs database directly.
+    global _MANAGED_JOBS_STALL_COLLECTOR
+    _MANAGED_JOBS_STALL_COLLECTOR = _wrap_collector(ManagedJobsStallCollector())
+    try:
+        prom.REGISTRY.register(_MANAGED_JOBS_STALL_COLLECTOR)
```
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply sky/server/metrics.py:1388]: Confirmed and fixed in `21cda58af`, and it was worse than a missing registration — thank you.

Verified the mode first rather than assuming: 15 processes in the api-server container on a real tenant carry `PROMETHEUS_MULTIPROC_DIR`, so multiprocess is the deployed shape, not an edge case. Every stall gauge would have been present in tests and absent in production, with `/metrics` still answering 200. Nothing downstream would have said so either: the alert that exists to catch "not measured" reads one of the missing families itself.

Two tests now cover it, because the reason the suite missed this is structural — the unit tests call `collect()` on the collector, and the list lives at the endpoint:

- one asserting this collector reaches the multiprocess registry;
- one general, comparing every collector that went through `_wrap_collector` against what that branch actually registers. Before the fix it named exactly one collector: this one.

Both were mutation-checked by restoring the branch without the registration; both go red.

File: sky/server/metrics.py [hunk 2]
```diff
@@ -931,6 +931,154 @@ def collect(self):
...
+        self._cache_ttl = _COLLECTOR_CACHE_TTL_SECONDS
+        # phase -> (tasks, scanned_at, truncated). A phase absent from this
+        # dict has never completed a scan and is exported as nothing at all.
+        # A phase whose refresh fails keeps its last good entry, and therefore
+        # its old timestamp, which is what the staleness alert reads.
+        self._cache: dict = {}
+
+    def _refresh(self):
+        # pylint: disable=import-outside-toplevel
+        from sky.jobs import stall
```
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply sky/server/metrics.py]: Correct, and the import is now at module scope in `21cda58af`.

It was copied from the collector next to it, which does use function-scope imports — but that is a habit, not a justification, and the guidance is explicit that a cycle has to force it.

Checked rather than assumed there is no cycle: with `from sky.jobs import stall` in the normal import block, all four orders import cleanly — `sky.server.metrics` first, `sky` first, `sky.server.server` first, and `sky.jobs.stall` first. So there was nothing to document, only an import to move.

File: sky/jobs/stall.py [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+    None -- with nothing contending it is MIN_PRIORITY, which suppresses
+    nothing. A failure here raises rather than being swallowed: this scan feeds
+    an alert, and the caller reports a scan that could not run as *not
+    measured* instead of as zero.
+    """
+    with _bounded(engine, deadline) as conn:
+        return managed_job_state.get_managed_jobs_highest_priority(conn)
+
+
+def _starved(task: StalledTask, highest: int) -> bool:
```
Comment (aylei) [inline sky/jobs/stall.py:497]: `never_claimed` reports every job queued behind the controller's own concurrency limit.

Repro: 5 RUNNING jobs, 50 queued at the same default priority for 20 min → 50 of 50 reported.

**Review by aylei — APPROVED** [review]
LGTM except the comment, thanks @DanielZhangQD !

File: sky/jobs/stall.py (hunk 3 above)
Comment (DanielZhangQD) [reply sky/jobs/stall.py:497]: Fixed
